crates/bevy_geometry/src/lib.rs

```diff
@@ -0,0 +1,359 @@
+    /// Returns true if this primitive is entirely on the outside (in the normal direction) of the
+    /// supplied plane.
+    fn outside_plane(&self, plane: Plane) -> bool;
+}
+
+#[derive(Debug, Clone)]
+pub enum PrimitiveError {
+    MinGreaterThanMax,
+    NonPositiveExtents,
+}
+impl Error for PrimitiveError {}
+impl fmt::Display for PrimitiveError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            PrimitiveError::MinGreaterThanMax => {
+                write!(
+                    f,
+                    "AxisAlignedBox minimums must be smaller or equal to the maximums"
+                )
+            }
+            PrimitiveError::NonPositiveExtents => {
+                write!(f, "AxisAlignedBox extents must be greater than zero")
+            }
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Debug, Reflect)]
+pub struct Sphere {
+    origin: Vec3,
+    radius: f32,
+}
+
+impl Sphere {
+    /// Get a reference to the sphere's origin.
+    pub fn origin(&self) -> &Vec3 {
+        &self.origin
+    }
+
+    /// Get a reference to the sphere's radius.
+    pub fn radius(&self) -> &f32 {
+        &self.radius
+    }
+
+    /// Set the sphere's origin.
+    pub fn set_origin(&mut self, origin: Vec3) {
+        self.origin = origin;
+    }
+
+    /// Set the sphere's radius.
+    pub fn set_radius(&mut self, radius: f32) {
+        self.radius = radius;
+    }
+}
+impl Primitive3d for Sphere {
+    /// Use the sphere's position and radius to determin eif it is entirely on the outside of the
+    /// the supplied plane.
+    fn outside_plane(&self, plane: Plane) -> bool {
+        plane.distance_to_point(&self.origin) > self.radius
+    }
+}
+
+/// An oriented box, unlike an axis aligned box, can be rotated and is not constrained to match the
+/// orientation of the coordinate system it is defined in. Internally, this is represented as an
+/// axis aligned box with some rotation ([Quat]) applied.
+#[derive(Copy, Clone, PartialEq, Debug, Reflect)]
+pub struct OrientedBox {
+    aab: AxisAlignedBox,
+    transform: Mat4,
+}
+impl Primitive3d for OrientedBox {
+    fn outside_plane(&self, plane: Plane) -> bool {
+        for vertex in self.vertices().iter() {
+            if plane.distance_to_point(vertex) <= 0.0 {
+                return false;
+            }
+        }
+        true
+    }
+}
+impl OrientedBox {
+    /// An ordered list of the vertices that form the 8 corners of the [AxisAlignedBox].
+    /// ```none
+    ///     (5)------(1)
+    ///      | \      | \
+    ///      |  (4)------(0)
+    ///      |   |    |   |
+    ///     (7)--|---(3)  |
+    ///        \ |      \ |
+    ///         (6)------(2)
+    /// ```
+    pub fn vertices(&self) -> [Vec3; 8] {
+        let mut vertices = [Vec3::ZERO; 8];
+        let aab_vertices = self.aab.vertices();
+        for i in 0..vertices.len() {
+            vertices[i] = self.transform.project_point3(aab_vertices[i])
```

Comment:
`project_point3` performs a perspective divide which is only necessary if the transform is not a perspective projection. If the transform was a perspective projection the the vertices would no longer be a box shape, so perhaps it is OK to assume that it is not. For an affine transform `transform_point3` is more efficient as it won't do an unnecessary divide.